kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt

```diff
@@ -1686,7 +1686,12 @@ internal class CodeGeneratorVisitor(
                 null
         )
 
-        return if (dstClass.isObjCClass()) {
+        return if (dstClass.isCompanion) {
+            call(
+                    llvm.Kotlin_Interop_IsSameClass,
+                    listOf(objCObject, genGetObjCClass(dstClass))
+            )
```

Comment:
If use pointer equality for type-checking, the above testcases print.
```kotlin
val nsObjectClass: ObjCClass = NSObject.Companion
// false
println(nsObjectClass is NSObject.Companion)

val nsObjectCompanion = NSObject.Companion
// true
println(nsObjectCompanion is NSObject.Companion)

// true
println(nsObjectMetaClass is NSObject.Companion)

// false
println(nsObjectClass == nsObjectMetaClass)
```

For me, the first case is confusing, it's so different with `C-family` languages or `Swift`.

Reply:
`functionGenerationContext.icmpEq(objCObject, genGetObjCClass(dstClass.parentAsClass))` is right.

Reply:
> That doesn't seem entirely right.
> 
> ```kotlin
> import kotlinx.cinterop.*
> import platform.darwin.*
> import platform.objc.*
> 
> @OptIn(kotlinx.cinterop.BetaInteropApi::class)
> fun main() {
>     val nsObjectClass: ObjCClass = NSObject
>     val nsObjectMetaClass: ObjCClass = object_getClass(nsObjectClass)!!
>     println(nsObjectClass is NSObject.Companion)
>     println(nsObjectMetaClass is NSObject.Companion)
>     println(nsObjectClass == nsObjectMetaClass)
> }
> ```
> 
> For me it prints
> 
> ```
> true
> true
> false
> ```
> 
> Meaning that there are two different objects, but both are instances of `NSObject.Companion`.
> 
> It might be more straightforward and correct to compare pointer equality when type-checking against such a companion class. What do you think?

`functionGenerationContext.icmpEq(objCObject, genGetObjCClass(dstClass.parentAsClass))` is the right thing.
The above testcase will print
```
val nsObjectClass: ObjCClass = NSObject
val nsObjectMetaClass: ObjCClass = object_getClass(nsObjectClass)!!

// true
println(nsObjectClass is NSObject.Companion)

// false
println(nsObjectMetaClass is NSObject.Companion)

// false
println(nsObjectClass == nsObjectMetaClass)
```